### PR TITLE
Implement missing parts for Windows. Change SO_REUSEPORT to SO_REUSEA…

### PR DESCRIPTION
**Pull request by beckdave**
…DDR, because SO_REUSEPORT is not available for Windows. I hope, it is the same behavior.

### CODE REVIEW

**Comment by lganzzzo** [comment]
Hey @beckdave ,

This looks great. Lets have Benedikt's input on this.

@bamkrs

File: src/oatpp-ssdp/SimpleUdpStreamProvider.cpp [hunk 1]
```diff
@@ -61,7 +62,55 @@ SimpleUdpStreamProvider::SimpleUdpStreamProvider(v_uint16 port) : m_port(port),
...
+    serverHandle = socket(result->ai_family, result->ai_socktype, result->ai_protocol);
+    if (serverHandle < 0) {
+        OATPP_LOGE("[oatpp::ssdp::SimpleUdpStreamProvider::instantiateServer()]", "Error. Couldn't open a socket: socket(%d, %d, %d) %s",
+            result->ai_family, result->ai_socktype, result->ai_protocol, strerror(errno));
+        throw std::runtime_error("[oatpp::ssdp::SimpleUdpStreamProvider::instantiateServer()]: Error. Couldn't open a socket");
+    }
+
+    m_closed = false;
+
+    ret = setsockopt(serverHandle, SOL_SOCKET, SO_REUSEADDR, (const char*)&yes, sizeof(int));
```
Comment (bamkrs) [inline src/oatpp-ssdp/SimpleUdpStreamProvider.cpp:95]: AFAIK for Windows, one has to set `SO_BROADCAST` for true multi-process compatibility. Without it, the applications would reuse the port, but the already running application won't receive packets anymore.

**Review by bamkrs — CHANGES_REQUESTED** [review]
Thank you for your PR! Windows support is always appreciated.

Could you please add `SO_BROADCAST` to the socket as well? This seems to be a best practice for true broadcast compatibility in Windows.

**Comment by beckdave** [comment]
Thank you for your answers. I was looking for some information about SO_REUSEPORT. I found this text on [https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ](https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ)

"The meaning of SO_REUSEADDR changes for multicast addresses as it allows multiple sockets to be bound to exactly the same combination of source multicast address and port. In other words, for multicast addresses SO_REUSEADDR behaves exactly as SO_REUSEPORT for unicast addresses. Actually, the code treats SO_REUSEADDR and SO_REUSEPORT identically for multicast addresses, that means you could say that SO_REUSEADDR implies SO_REUSEPORT for all multicast addresses and the other way round."

For me, it sounds like changing SO_REUSEPORT to SO_REUSEADDR has no difference in case of multicast.

What do you think?

**Comment by bamkrs** [comment]
Its not about `SO_REUSEPORT` to `SO_REUSEADDR`. Its about adding `SO_BROADCAST` as another socketop for Windows. Albeit its only for sending, one may end up in the situation of sending broadcasts with this socket. And since we are aiming for true Windows-compatibility, this is a necessary socketop.

**Comment by beckdave** [comment]
Thank you for the Information. I will add it, only for Windows.

**Comment by beckdave** [comment]
Some Questions:
- The SO_REUSEADDR socket option is already set in SimpleUdpStreamProvider. I think, it is not necessary to set it again in SimpleSsdpUdpStreamProvider.
- Why is SO_BROADCAST needed for Windows, but not for Linux?

**Comment by bamkrs** [comment]
> * Why is SO_BROADCAST needed for Windows, but not for Linux?

Oh dang! TIL it's also a flag for Linux. I've always read it in company with windows and must have missed it in Linux. You are right, `SO_BROADCAST` should also be set for Linux!

**Comment by beckdave** [comment]
I pushed a commit:
- I removed SO_REUSEADDR in SimpleSsdpUdpStreamProvider, because it is already set in SimpleUdpStreamProvider. I think SO_REUSEADDR shall even be set before bind(). But in SimpleSsdpUdpStreamProvider it is set after bind(). So, the right place to set SO_REUSEADDR is SimpleUdpStreamProvider.
- I enabled SO_BROADCAST for all plattforms.

**Comment by bamkrs** [comment]
I totally agree, nice catch. Good work!
The code was refactored and rewritten multiple times because @lganzzzo and I tried so many different things how to implement a non-stream based protocol in Oat++ which is entirely build upon streams. So its nice to have another pair of eyes looking through it and fix stuff we missed because we were obsessed with other details.

**Comment by beckdave** [comment]
Thank you for your comment. I appreciated to improve oatpp-ssdp a little bit. Thank you very much for oatpp, it is a great lib.
If you are interested in my communication lib, just check it out: [https://github.com/bexoft/finalmq](https://github.com/bexoft/finalmq).

**Pull request merged**